middle_end/flambda/inlining/inlining_transforms.ml

```diff
@@ -23,6 +23,169 @@ module DE = Downwards_env
+     [let_cont_exn k1 (exn: val) =
+       Apply_cont k_exn exn extra_args
+     in
+     let_cont k_pop (args) = Apply_cont<pop k1> k args in
+     let_cont k_push () = Apply_expr f (args) <k_pop> «k1» in
+     Apply_cont<push k1> k_push ()]
+
+     This feels quite heavy, but is necessary because we can rewrite
+     neither the definition and other uses of k_exn nor the
+     uses of the exception continuation in the body of f, so we need
+     two distinct exception continuations; and of course the new
+     exception continuation needs to be correctly pushed and popped.
+
+     The most annoying part of this is that it introduces trywith
+     blocks that were not part of the initial program, will not be
+     removed, and might be useless (if the function never raises).
+
+     Maybe a better solution would be to propagate through dacc
+     a lazy rewriting, that would add the correct extra args
+     to all uses of the exception continuation in the body.
+   *)
+  let wrapper = Continuation.create ~sort:Exn () in
+  let body_with_pop =
+    match (apply_return_continuation : Apply.Result_continuation.t) with
+    | Never_returns ->
+       make_inlined_body ~apply_exn_continuation:wrapper
+         ~apply_return_continuation
+    | Return apply_return_continuation ->
+       let pop_wrapper_cont = Continuation.create () in
+       let pop_wrapper_handler =
+         let kinded_params =
+           List.map (fun k -> Variable.create "wrapper_return", k)
+             result_arity
+         in
+         let trap_action =
+           Trap_action.Pop { exn_handler = wrapper; raise_kind = None; }
+         in
+         let args = List.map (fun (v, _) -> Simple.var v) kinded_params in
+         let apply_cont =
+           Apply_cont.create ~trap_action
+             apply_return_continuation
+             ~args
+             ~dbg:Debuginfo.none
+         in
+         let handler = Expr.create_apply_cont apply_cont in
+         Continuation_handler.create
+           (Kinded_parameter.List.create kinded_params)
+           ~handler
+           ~free_names_of_handler:Unknown
+           ~is_exn_handler:false
+       in
+       let new_apply_return_continuation =
+         Apply.Result_continuation.Return pop_wrapper_cont
+       in
+       let body =
+         make_inlined_body ~apply_exn_continuation:wrapper
+           ~apply_return_continuation:new_apply_return_continuation
+       in
+       Let_cont.create_non_recursive pop_wrapper_cont
+         pop_wrapper_handler ~body
+         ~free_names_of_body:Unknown
+  in
+  let wrapper_handler =
+    let param = Variable.create "exn" in
+    let kinded_params =
+      [KP.create param K.With_subkind.any_value]
+    in
+    let exn_handler =
+      Exn_continuation.exn_handler apply_exn_continuation
+    in
+    let trap_action =
+      Trap_action.Pop { exn_handler; raise_kind = None; }
+    in
+    let handler =
+      Expr.create_apply_cont (Apply_cont.create
```

Comment:
This indentation is unpleasant, I assume an ocamlformat config needs updating.  In the meantime maybe this could be fixed by doing `Apply_cont.create` then `|> Expr.create_apply_cont` at the end.  I've been adopting this style elsewhere in any case.

Reply:
Done